pytorch_lightning/callbacks/pruning.py

```diff
@@ -0,0 +1,316 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+r"""
+Pruning
+^^^^^^^
+
+Perform model pruning.
+
+"""
+from copy import deepcopy
+from typing import Callable, List, Optional, Tuple, Union
+
+from torch import nn
+from torch.nn.modules.container import ModuleDict, ModuleList
+
+from pytorch_lightning.callbacks.base import Callback
+from pytorch_lightning.core.lightning import LightningModule
+from pytorch_lightning.utilities import _PYTORCH_PRUNE_AVAILABLE
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+if _PYTORCH_PRUNE_AVAILABLE:
+    import torch.nn.utils.prune as pytorch_prune
+
+
+_PYTORCH_PRUNING_FUNCTIONS = {
+    "ln_structured": pytorch_prune.ln_structured,
+    "l1_unstructured": pytorch_prune.l1_unstructured,
+    "random_structured": pytorch_prune.random_structured,
+    "random_unstructured": pytorch_prune.random_unstructured,
+}
+
+_PYTORCH_PRUNING_METHOD = {
+    "ln_structured": pytorch_prune.LnStructured,
+    "l1_unstructured": pytorch_prune.L1Unstructured,
+    "random_structured": pytorch_prune.RandomStructured,
+    "random_unstructured": pytorch_prune.RandomUnstructured,
+}
+
+
+def check_parameters_to_prune(
+    pl_module: LightningModule,
+    parameters_to_prune: Optional[List],
+    parameters: List[str] = ["weight"]
+) -> List:
+
+    is_parameters_to_prune_none = parameters_to_prune is None
+    current_modules = [m for m in pl_module.modules() if not isinstance(m, (LightningModule, ModuleDict, ModuleList))]
+
+    if is_parameters_to_prune_none:
+        parameters_to_prune = []
+        for p in parameters:
+            for m in current_modules:
+                param = getattr(m, p, None)
+                if param is not None:
+                    parameters_to_prune.append((m, p))
+
+    if isinstance(parameters_to_prune, (tuple, list)) \
+       and len(parameters_to_prune) > 0 and not is_parameters_to_prune_none:
+
+        if all(isinstance(p, (list, tuple)) and len(p) == 2 for p in parameters_to_prune):
```

Comment:
This check seems very elusive and I don't like a magic number in the call here (`2`). Can we clear this up? It sees we append the tuples into `parameters_to_prune` so not sure if this check is needed

Reply:
Done !